### DISCUSSION TITLE
Error calculating second-order information

### DISCUSSION THREAD

**Discussion by WouterMuijres** [casadi/casadi#3226] [Q&A]
Dear CasADi team and users,

I am running into problems when trying to calculate second-order derivatives from an external function (in MATLAB). 

I can calculate the Jacobian of the external function. However, when I structure the expression of Jacobian into a vector, I cannot calculate the Jacobian a second time to obtain an expression for the second-order information (see error below). The zipped folder contains the scripts that reproduce the error (Sorry, I was not able to add the .dll file to GitHub gist).

```
Error:

Error using casadi.GenericMatrixCommon/jacobian
Error in MX::jacobian at .../casadi/core/mx.cpp:1708:
Error in XFunction::jac for 'helper_jacobian_MX' [MXFunction] at .../casadi/core/x_function.hpp:804:
Error in MXFunction::ad_forward at .../casadi/core/mx_function.cpp:890:
[Question2ndOrderInfoCasadiForum.zip](https://github.com/casadi/casadi/files/11721733/Question2ndOrderInfoCasadiForum.zip)

Error in MX::ad_forward for node of type N6casadi4CallE at .../casadi/core/mx.cpp:2301:
Error in Call::ad_forward for 'fwd5_F' [MXFunction] at .../casadi/core/casadi_call.cpp:121:
Error in Function::forward for 'fwd5_F' [MXFunction] at .../casadi/core/function.cpp:1083:
Error in XFunction::get_forward for 'fwd5_F' [MXFunction] at .../casadi/core/x_function.hpp:856:
Error in MXFunction::ad_forward at .../casadi/core/mx_function.cpp:890:
Error in MX::ad_forward for node of type N6casadi4CallE at .../casadi/core/mx.cpp:2301:
Error in Call::ad_forward for 'map5_fwd1_F' [Map] at .../casadi/core/casadi_call.cpp:121:
Error in Function::forward for 'map5_fwd1_F' [Map] at .../casadi/core/function.cpp:1083:
Error in Function::forward for 'fwd1_F' [External] at .../casadi/core/function.cpp:1083:

Error in TestJacJac (line 35)
JF2         = jacobian(JF_vertcat,q_MX);
```

I am able to generate an expression for second-order information by only selecting two outputs from the external function’s Jacobian (see example script). 

In my application, I use the Jacobian of the dynamics to constrain the problem. This results in an error (above) when calculating the constraint Jacobian. 

The external function describes the dynamics of a 5 DOF conceptual model. We used MATLAB code to generate the DLL file (Github: https://github.com/Lars-DHondt-KUL/opensimAD).

Would you be able to guide me in the direction of a solution? I would be happy to send you additional scripts/functions which may help you to understand the problem better.

Thanks in advance.

Kind regards,

Wouter Muijres
PhD-student 
Human Movement Biomechanics Research Group
[Question2ndOrderInfoCasadiForum.zip](https://github.com/casadi/casadi/files/11721761/Question2ndOrderInfoCasadiForum.zip)

**Comment by jaeandersson** [comment]
Since it involves external functions, debugging gets a bit more involved. I would have to look into the code and run it to point to understand what's happening, and I don't have resources to do that for questions on the open forum. Since both you and @jgillis are at KUL, maybe he can help you further.
